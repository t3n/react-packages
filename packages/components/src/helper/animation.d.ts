import { ThemeProps } from '@t3n/theme';
export declare const backgroundAnimation: import("styled-components").FlattenInterpolation<import("styled-components").ThemedStyledProps<ThemeProps, any>>;
