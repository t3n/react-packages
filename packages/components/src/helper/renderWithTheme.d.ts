import 'jest-styled-components';
import React from 'react';
import { RenderOptions } from '@testing-library/react';
import { Theme } from '@t3n/theme';
interface OptionalTheme {
    theme?: Theme;
}
export declare const renderWithTheme: (ui: React.ReactElement, { theme, ...options }: RenderOptions & OptionalTheme) => import("@testing-library/react").RenderResult<typeof import("@testing-library/dom/types/queries"), HTMLElement>;
export {};
