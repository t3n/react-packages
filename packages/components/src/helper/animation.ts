import { css, keyframes } from 'styled-components';

import { ThemeProps } from '@t3n/theme';

const backgroundAnimationKeyframes = ({ theme }: ThemeProps) => keyframes`
  0% {
    background-color: ${theme.colors.shades.grey204}
  }

  50% {
    background-color: ${theme.colors.shades.grey244}
  }

  100% {
    background-color: ${theme.colors.shades.grey204}
  }
`;

// eslint-disable-next-line import/prefer-default-export
export const backgroundAnimation = css`
  animation: ${backgroundAnimationKeyframes} 2s linear infinite;
`;
