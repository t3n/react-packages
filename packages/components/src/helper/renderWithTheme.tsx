/* eslint-disable import/no-extraneous-dependencies */
import 'jest-styled-components';

import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Theme, theme as originalTheme } from '@t3n/theme';

interface OptionalTheme {
  theme?: Theme;
}

// eslint-disable-next-line import/prefer-default-export
export const renderWithTheme = (
  ui: React.ReactElement,
  { theme, ...options }: RenderOptions & OptionalTheme
) => {
  // merge over default theme

  const mergedTheme = { ...originalTheme, ...theme };

  const Wrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={mergedTheme}>{children}</ThemeProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};
