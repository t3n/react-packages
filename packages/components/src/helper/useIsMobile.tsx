import { Theme } from '@t3n/theme';
import { useCallback, useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';

const useIsMobile = (): boolean => {
  const [isMobile, setIsMobile] = useState(false);

  const theme: Theme = useContext(ThemeContext);
  const fontSize = window
    .getComputedStyle(window.document.body)
    .getPropertyValue('font-size');
  const mobileBreakpoint =
    parseInt(theme.breakpoints[1], 10) * parseInt(fontSize, 10);

  const testIsMobile = useCallback(() => {
    if (window.matchMedia(`(max-width: ${mobileBreakpoint}px)`).matches) {
      setIsMobile(true);
    } else {
      setIsMobile(false);
    }
  }, [mobileBreakpoint]);

  useEffect(() => {
    testIsMobile();
    window.addEventListener('resize', testIsMobile);

    return () => {
      window.removeEventListener('resize', testIsMobile);
    };
  }, [testIsMobile]);

  return isMobile;
};

export default useIsMobile;
