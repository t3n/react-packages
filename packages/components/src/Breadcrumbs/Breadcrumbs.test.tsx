import React from 'react';

import { renderWithTheme } from '../helper/renderWithTheme';
import {
  Breadcrumbs,
  BreadcrumbsItem,
  BreadcrumbsItemProps,
} from './Breadcrumbs';

const renderBreadcrumbs = () =>
  renderWithTheme(
    <Breadcrumbs>
      <BreadcrumbsItem label="Item 1" href="/item1" />
      <BreadcrumbsItem label="Item 2" href="/item2" />
      <BreadcrumbsItem label="Item 3" />
    </Breadcrumbs>,
    {}
  );

test('Breadcrumbs matches snapshot', () => {
  const { container } = renderBreadcrumbs();

  expect(container.firstChild).toMatchSnapshot();
});

test('Breadcrumbs navigation element is accessible', () => {
  const { getByRole } = renderBreadcrumbs();

  expect(getByRole('navigation').getAttribute('aria-label')).toEqual(
    'Breadcrumb'
  );
});

test('BreadcrumbsItems render label prop as text content', () => {
  const { getByText } = renderBreadcrumbs();

  expect(getByText('Item 1')).toBeDefined();
  expect(getByText('Item 2')).toBeDefined();
  expect(getByText('Item 3')).toBeDefined();
});

test('BreadcrumbsItems with href prop provided render as anchor element', () => {
  const { getByText } = renderBreadcrumbs();

  expect(getByText('Item 1').nodeName.toLowerCase()).toEqual('a');
  expect(getByText('Item 2').nodeName.toLowerCase()).toEqual('a');
  expect(getByText('Item 3').nodeName.toLowerCase()).not.toEqual('a');

  expect(getByText('Item 1').getAttribute('href')).toEqual('/item1');
  expect(getByText('Item 2').getAttribute('href')).toEqual('/item2');
  expect(getByText('Item 3').getAttribute('href')).toBeFalsy();
});

test('BreadcrumbsItems without href prop provided render as span element', () => {
  const { getByText } = renderBreadcrumbs();

  expect(getByText('Item 1').nodeName.toLowerCase()).not.toEqual('span');
  expect(getByText('Item 2').nodeName.toLowerCase()).not.toEqual('span');
  expect(getByText('Item 3').nodeName.toLowerCase()).toEqual('span');
});

test('BreadcrumbsItems without href render with bold text style', () => {
  const { getByText } = renderBreadcrumbs();

  expect(window.getComputedStyle(getByText('Item 1')).fontWeight).not.toBe(
    'bold'
  );
  expect(window.getComputedStyle(getByText('Item 2')).fontWeight).not.toBe(
    'bold'
  );
  expect(window.getComputedStyle(getByText('Item 3')).fontWeight).toBe('bold');
});

test('BreadcrumbsItems render custom link component if provided through props', () => {
  const CustomLink = ({ href, title }: { href: string; title?: string }) => (
    <a href={href} title={title} style={{ color: 'red' }}>
      {title}
    </a>
  );

  const CustomBreadcrumbsItem = (props: BreadcrumbsItemProps) => (
    <BreadcrumbsItem {...props} linkComponent={CustomLink} />
  );

  const { container, getByText } = renderWithTheme(
    <Breadcrumbs>
      <CustomBreadcrumbsItem label="Item 1" href="/item1" />
      <CustomBreadcrumbsItem label="Item 2" href="/item2" />
      <CustomBreadcrumbsItem label="Item 3" />
    </Breadcrumbs>,
    {}
  );

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText('Item 1').style.color).toBe('red');
  expect(getByText('Item 2').style.color).toBe('red');
  expect(getByText('Item 3').style.color).not.toBe('red');
});
