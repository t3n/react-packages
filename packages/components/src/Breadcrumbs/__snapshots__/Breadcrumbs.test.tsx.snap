// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs matches snapshot 1`] = `
.c3 {
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  font-weight: bold;
}

.c2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  pointer-events: auto;
  opacity: 1;
  background-repeat: repeat-x;
  background-size: 1px 1px;
  background-position: 0 100%;
  cursor: pointer;
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  color: #2a2a2a;
  background-image: url('data:image/svg+xml;base64,PHN2ZyBwcmVzZXJ2ZUFzcGVjdFJhdGlvPSJub25lIiB2aWV3Qm94PSIwIDAgMSAxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxnIHN0cm9rZT0icmdiKDk1LDk1LDk1KSI+PHJlY3QgeD0iMCIgeT0iMCIgd2lkdGg9IjEiIGhlaWdodD0iMSIgLz48L2c+PC9zdmc+');
}

.c2:hover {
  color: #2a2a2a;
  background-image: url('data:image/svg+xml;base64,PHN2ZyBwcmVzZXJ2ZUFzcGVjdFJhdGlvPSJub25lIiB2aWV3Qm94PSIwIDAgMSAxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxnIHN0cm9rZT0icmdiKDI0OSw2Niw1OCkiPjxyZWN0IHg9IjAiIHk9IjAiIHdpZHRoPSIxIiBoZWlnaHQ9IjEiIC8+PC9nPjwvc3ZnPg==');
}

.c2:focus {
  color: #f9423a;
  background-image: url('data:image/svg+xml;base64,PHN2ZyBwcmVzZXJ2ZUFzcGVjdFJhdGlvPSJub25lIiB2aWV3Qm94PSIwIDAgMSAxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxnIHN0cm9rZT0icmdiKDI0OSw2Niw1OCkiPjxyZWN0IHg9IjAiIHk9IjAiIHdpZHRoPSIxIiBoZWlnaHQ9IjEiIC8+PC9nPjwvc3ZnPg==');
}

.c2:visited {
  color: #2a2a2a;
  background-image: url('data:image/svg+xml;base64,PHN2ZyBwcmVzZXJ2ZUFzcGVjdFJhdGlvPSJub25lIiB2aWV3Qm94PSIwIDAgMSAxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxnIHN0cm9rZT0icmdiKDk1LDk1LDk1KSI+PHJlY3QgeD0iMCIgeT0iMCIgd2lkdGg9IjEiIGhlaWdodD0iMSIgLz48L2c+PC9zdmc+');
}

.c2:visited:focus {
  color: #f9423a;
  background-image: url('data:image/svg+xml;base64,PHN2ZyBwcmVzZXJ2ZUFzcGVjdFJhdGlvPSJub25lIiB2aWV3Qm94PSIwIDAgMSAxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxnIHN0cm9rZT0icmdiKDI0OSw2Niw1OCkiPjxyZWN0IHg9IjAiIHk9IjAiIHdpZHRoPSIxIiBoZWlnaHQ9IjEiIC8+PC9nPjwvc3ZnPg==');
}

.c1 {
  position: relative;
  padding-left: 16px;
  padding-right: 16px;
  line-height: 1.666em;
  margin-right: -2px;
}

.c1:after {
  content: '';
  position: absolute;
  display: block;
  width: 0;
  height: 0;
  bottom: 50%;
  left: 0;
  margin-bottom: -6px;
  border-top: 0.333rem solid transparent;
  border-bottom: 0.333rem solid transparent;
  border-left: 0.333rem solid #5f5f5f;
}

.c1:first-child {
  padding-left: 0;
}

.c1:first-child:after {
  display: none;
}

.c1:last-child {
  padding-right: 0;
}

.c0 {
  margin: 0;
  padding: 0;
  padding-bottom: 1rem;
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

@media screen and (min-width:30rem) {
  .c3 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c3 {
    line-height: 1.666em;
  }
}

@media screen and (min-width:30rem) {
  .c2 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c2 {
    line-height: 1.666em;
  }
}

<nav
  aria-label="Breadcrumb"
>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <a
        class="c2"
        href="/item1"
        title="Item 1"
      >
        Item 1
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        class="c2"
        href="/item2"
        title="Item 2"
      >
        Item 2
      </a>
    </li>
    <li
      class="c1"
    >
      <span
        class="c3"
      >
        Item 3
      </span>
    </li>
  </ul>
</nav>
`;

exports[`BreadcrumbsItems render custom link component if provided through props 1`] = `
.c2 {
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  font-weight: bold;
}

.c1 {
  position: relative;
  padding-left: 16px;
  padding-right: 16px;
  line-height: 1.666em;
  margin-right: -2px;
}

.c1:after {
  content: '';
  position: absolute;
  display: block;
  width: 0;
  height: 0;
  bottom: 50%;
  left: 0;
  margin-bottom: -6px;
  border-top: 0.333rem solid transparent;
  border-bottom: 0.333rem solid transparent;
  border-left: 0.333rem solid #5f5f5f;
}

.c1:first-child {
  padding-left: 0;
}

.c1:first-child:after {
  display: none;
}

.c1:last-child {
  padding-right: 0;
}

.c0 {
  margin: 0;
  padding: 0;
  padding-bottom: 1rem;
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

@media screen and (min-width:30rem) {
  .c2 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c2 {
    line-height: 1.666em;
  }
}

<nav
  aria-label="Breadcrumb"
>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <a
        href="/item1"
        style="color: red;"
        title="Item 1"
      >
        Item 1
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        href="/item2"
        style="color: red;"
        title="Item 2"
      >
        Item 2
      </a>
    </li>
    <li
      class="c1"
    >
      <span
        class="c2"
      >
        Item 3
      </span>
    </li>
  </ul>
</nav>
`;
